// @curveball/controller/dist/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.controllerDecorator = exports.accept = exports.method = void 0;
function method(httpMethod) {
    return controllerDecorator('method', httpMethod);
}
exports.method = method;
function accept(mimeType) {
    return controllerDecorator('accept', mimeType);
}
exports.accept = accept;
/**
 * This function returns a very simple decorator.
 *
 * All it does is take the name of the annotation, the arguments that were
 * passed and add their contents to a 'annotations' property on the
 * controller.
 */
function controllerDecorator(annotationName, ...args) {
    return (target, propertyKey) => {
        if (!target.annotations) {
            target.annotations = new Map();
        }
        if (!target.annotations.has(propertyKey)) {
            target.annotations.set(propertyKey, []);
        }
        target.annotations.get(propertyKey).push({
            name: annotationName,
            args: args
        });
    };
}
exports.controllerDecorator = controllerDecorator;
